File: scripts/tide/sweep_process/collect.py
#!/usr/bin/env python3
"""collect.py — assemble per-job sweep CSVs into one long-format table.

Run after fetch.sh has pulled per-job outputs to ``RAW_DIR``. Each per-job
CSV is read, tagged with ``fire``, ``r_D``, ``r_S`` parsed from its
directory name, and concatenated into a single CSV at ``OUT_FILE``.

Usage (from the repo root, after fetch.sh):
    python scripts/tide/sweep_process/collect.py

Optional environment overrides:
    RAW_DIR   directory holding per-job subdirs (default: _data/processed/sweep)
    OUT_FILE  output CSV path (default: _data/processed/sweep/sweep_all.csv)
"""

from __future__ import annotations

import os
import re
from pathlib import Path

import pandas as pd

RAW_DIR = Path(os.environ.get("RAW_DIR", "_data/processed/sweep"))
OUT_FILE = Path(os.environ.get("OUT_FILE", "_data/processed/sweep/sweep_all.csv"))

NAME_RE = re.compile(r"^(?P<fire>[a-z]+)_rD(?P<r_D>\d+)_rS(?P<r_S>\d+)$")


def collect() -> pd.DataFrame:
    if not RAW_DIR.exists():
        raise SystemExit(f"RAW_DIR not found: {RAW_DIR}")

    frames = []
    for sub in sorted(RAW_DIR.iterdir()):
        if not sub.is_dir():
            continue
        m = NAME_RE.match(sub.name)
        if not m:
            print(f"  skipping {sub.name} (doesn't match <fire>_rD<n>_rS<n>)")
            continue
        csv = sub / f"{sub.name}_raw_metrics.csv"
        if not csv.exists():
            print(f"  WARNING: {csv} missing")
            continue
        df = pd.read_csv(csv)
        df.insert(0, "fire", m["fire"])
        df.insert(1, "r_D", int(m["r_D"]))
        df.insert(2, "r_S", int(m["r_S"]))
        frames.append(df)
        print(f"  + {sub.name}: {len(df):,} rows")

    if not frames:
        raise SystemExit(f"No matching per-job CSVs found under {RAW_DIR}")

    return pd.concat(frames, ignore_index=True)


def main() -> None:
    print(f"Collecting sweep outputs from: {RAW_DIR}")
    out = collect()
    OUT_FILE.parent.mkdir(parents=True, exist_ok=True)
    out.to_csv(OUT_FILE, index=False)
    print(f"\nWrote {OUT_FILE}  ({len(out):,} rows, {len(out.columns)} cols)")
    print("Combinations present:")
    print(
        out.groupby(["fire", "r_D", "r_S"])
        .size()
        .rename("n_rows")
        .reset_index()
        .to_string(index=False)
    )


if __name__ == "__main__":
    main()

File: scripts/tide/sweep_cv/collect.R
#!/usr/bin/env Rscript
## collect.R — assemble per-job CV CSVs into one results table, plus a
## per-cell summary and a top-forms "optimum" table.
##
## Run after fetch.sh has pulled per-job outputs to RAW_DIR.
##
## Usage (from the repo root, after fetch):
##   Rscript scripts/tide/sweep_cv/collect.R
##
## Optional env overrides:
##   RAW_DIR       directory holding per-job subdirs (default _data/processed/sweep_cv)
##   OUT_RESULTS   long-format CSV with every fold's score (default <RAW_DIR>/sweep_results.csv)
##   OUT_SUMMARY   mean-per-setting summary (default <RAW_DIR>/sweep_summary.csv)
##   OUT_OPTIMA    top forms per scale (default <RAW_DIR>/sweep_optima.csv)

suppressPackageStartupMessages({
  library(dplyr)
  library(readr)
  library(purrr)
})

RAW_DIR     <- Sys.getenv("RAW_DIR",     "_data/processed/sweep_cv")
OUT_RESULTS <- Sys.getenv("OUT_RESULTS", file.path(RAW_DIR, "sweep_results.csv"))
OUT_SUMMARY <- Sys.getenv("OUT_SUMMARY", file.path(RAW_DIR, "sweep_summary.csv"))
OUT_OPTIMA  <- Sys.getenv("OUT_OPTIMA",  file.path(RAW_DIR, "sweep_optima.csv"))

# ----- Find per-job CSVs ------------------------------------------------------

if (!dir.exists(RAW_DIR)) stop("RAW_DIR not found: ", RAW_DIR)
files <- list.files(RAW_DIR, pattern = "_cv\\.csv$", recursive = TRUE, full.names = TRUE)
if (length(files) == 0L) stop("No per-job *_cv.csv files under ", RAW_DIR)

message(sprintf("Found %d per-job CV CSVs", length(files)))

# ----- Concatenate ------------------------------------------------------------

results <- map_dfr(files, read_csv, show_col_types = FALSE)
message(sprintf("Concatenated to %d rows", nrow(results)))

dir.create(dirname(OUT_RESULTS), recursive = TRUE, showWarnings = FALSE)
write_csv(results, OUT_RESULTS)
message(sprintf("Wrote %s", OUT_RESULTS))

# ----- Per-cell summary (mean over folds) ------------------------------------
# One row per (r_D, scale, cv, form, fire). score = AUC (structure) /
# concordance (neighbourhood); aux = log_loss / rmse.

summary_tbl <- results |>
  group_by(r_D, scale, cv, sd_form, ss_form, form, fire) |>
  summarise(
    n_folds    = dplyr::n(),
    score_mean = mean(score, na.rm = TRUE),
    score_sd   = sd  (score, na.rm = TRUE),
    aux_mean   = mean(aux,   na.rm = TRUE),
    .groups = "drop"
  ) |>
  arrange(scale, cv, fire, form, r_D)

write_csv(summary_tbl, OUT_SUMMARY)
message(sprintf("Wrote %s", OUT_SUMMARY))

# ----- Optima: best (r_D, form) per scale, on the pooled held-out set --------

optima <- summary_tbl |>
  filter(cv == "pooled", fire == "pooled") |>
  group_by(scale) |>
  slice_max(score_mean, n = 5L, with_ties = FALSE) |>
  ungroup() |>
  arrange(scale, desc(score_mean))

write_csv(optima, OUT_OPTIMA)
message(sprintf("Wrote %s", OUT_OPTIMA))

# ----- Print to console ------------------------------------------------------

cat("\n=== Top (r_D, form) per scale on the pooled held-out set ===\n")
print(optima, n = Inf, width = Inf)

cat("\n=== Shipped form (SD_uniform_root_area | SS_flat), by scale/fire/r_D ===\n")
summary_tbl |>
  filter(cv == "pooled", form == "SD_uniform_root_area|SS_flat") |>
  arrange(scale, fire, r_D) |>
  print(n = Inf, width = Inf)
